addons/document/document.py

```diff
@@ -108,12 +108,27 @@ def search(self, cr, uid, args, offset=0, limit=None, order=None, context=None,
         for parent_id in visible_parent_ids:
             ids.extend(parents[parent_id])
 
+        set_ids = set(ids)
+
         # sort result according to the original sort ordering
         if count:
-            return len(ids)
+            result = len(set_ids)
         else:
-            set_ids = set(ids)
-            return [id for id in orig_ids if id in set_ids]
+            result = [id for id in orig_ids if id in set_ids]
+
+        # if we got a limit and some ids were removed, search again
+        if limit and (len(set_ids) < len(orig_ids) or not orig_ids):
+            # make sure we don't try to search more records than exist
+            if not orig_ids and offset > self.search(
```

Comment:
I'm not sure about needing this condition, as the following search will return [], and thus the result will be the same.